google-cloud-logging/lib/google/cloud/logging/async_writer.rb

```diff
@@ -201,182 +197,377 @@ def logger log_name, resource, labels = {}
+
+              if @batch.ready?
+                # interval met, publish the batch...
+                publish_batch!
+                @cond.wait
+              else
+                # still waiting for the interval to publish the batch...
+                @cond.wait(@batch.publish_wait)
+              end
+            end
           end
         end
 
-        protected
+        def publish_batch!
+          return unless @batch
 
-        ##
-        # @private The background thread implementation, which continuously
-        # waits for and performs work, and returns only when fully stopped.
-        #
-        def run_backgrounder
-          queue_item = wait_next_item
-          return unless queue_item
-          begin
-            logging.write_entries(
-              queue_item.entries,
-              log_name: queue_item.log_name,
-              resource: queue_item.resource,
-              labels: queue_item.labels,
-              partial_success: @partial_success
-            )
-          rescue StandardError => e
-            # Ignore any exceptions thrown from the background thread, but
-            # keep running to ensure its state behavior remains consistent.
-            @last_exception = e
+          publish_batch_async @batch
+          @batch = nil
+        end
+
+        # Sets the last_error and calls all error callbacks.
+        def error! error
+          error_callbacks = synchronize do
+            @last_error = error
+            @error_callbacks
           end
+          error_callbacks.each { |error_callback| error_callback.call error }
         end
 
-        ##
-        # @private Wait for and dequeue the next set of log entries to transmit.
-        #
-        # @return [QueueItem, nil] Returns the next set of entries. If
-        #   the writer has been stopped and no more entries are left in the
-        #   queue, returns `nil`.
-        #
-        def wait_next_item
-          synchronize do
-            while state == :suspended ||
-                  (state == :running && @queue.empty?)
-              @queue_resource.wait
+        def publish_batch_async batch
+          batch.values.each do |write_log_entries_request|
+            begin
+              write_entries_future write_log_entries_request
+            rescue Concurrent::RejectedExecutionError => e
+              async_error = AsyncWriterError.new(
+                "Error writing entries: #{e.message}",
+                write_log_entries_request.entries
+              )
+              # Manually set backtrace so we don't have to raise
+              async_error.set_backtrace backtrace
+              error! async_error
             end
-            queue_item = nil
-            unless @queue.empty?
-              queue_item = @queue.shift
-              @queue_size -= queue_item.entries.size
+          end
+        end
+
+        def write_entries_future write_log_entries_request
+          Concurrent::Future.execute(executor: @thread_pool) do
```

Comment:
Just a design nit: I'd consider putting the `Concurrent::Future.execute` call itself in publish_batch_async because that's where RejectedExecutionError is being caught. Then this method can just include the code in the block: the API call itself along with its rescue of StandardError. That puts all the concurrent-ruby stuff in one place and the API stuff together in another place.

Reply:
Also, should we consider using [`Concurrent::Promises`](http://ruby-concurrency.github.io/concurrent-ruby/master/Concurrent/Promises.html) instead of `Concurrent::Future` now that the former is stable?

Reply:
There are two different situations that need to be handled, and they cannot be handled with a single `rescue`. The first `rescue` in `publish_batch_async` is for an error raised when executing the future on a thread pool. This can happen when the thread pool's queue has reached the limit set.

The second `rescue` in `write_entries_future` is for an error raised within the running future, while making the API call. Both situations need to be handled.

Originally the logic in `publish_batch_async` and `write_entries_future` were in a single method, but were split up into separate methods is to satisfy Rubocop.

Reply:
I think a rewrite to use `Concurrent::Promises` should be a separate PR. I would prefer to see separate competing PRs to making that change in this PR.

Reply:
Regarding `Concurrent::Promises`, sure, we can defer that. I think it's a trivial change; I'll open it as a separate PR once we merge this one.

Regarding the rescues, yes, I get that there need to be two separate rescues. What I'm suggesting is lifting _only_ the `Concurrent::Future.execute(executor: @thread_pool) do` line and its `end` up into the `publish_batch_async` method. So it looks like:

```ruby
  begin
    Concurrent::Future.execute(executor: @thread_pool) do
      write_entries_future write_log_entries_request
    end
  rescue Concurrent::RejectedExecutionError => e
    # ...
```

That way, `publish_batch_async` contains the call to `Concurrent::Future.execute` _and_ handles the `RejectedExecutionError` that it may raise. And `write_entries_future` contains the call to `logging.write_entries` and handles the errors that _it_ may raise. In other words `publish_batch_async` handles only the concerns of concurrency (including its error handling), and `write_entries_future` handles only the concerns of the logging API (including its error handling).

Reply:
Updated. PTAL.